cmake/GenerateBuiltins.cmake

```diff
@@ -503,3 +503,125 @@ function (generate_builtins)
         ${DISPATCH_BUILTIN_CPP_FILES} ${COMMON_BUILTIN_CPP_FILES} ${TARGET_BUILTIN_CPP_FILES})
     add_dependencies(builtin builtins-cpp)
 endfunction()
+
+function (generate_generic_target_builtin_1 ispc_name target arch bit os)
+    set(include ${CMAKE_CURRENT_SOURCE_DIR}/stdlib/include)
+
+    set(name "builtins_target_generic_${target}_${arch}_${os}")
+    set(input_ispc builtins/generic.ispc)
+
+    if ("${os}" STREQUAL "unix")
+        set(fixed_os "linux")
+        set(OS_UP UNIX)
+    elseif ("${os}" STREQUAL "windows")
+        set(fixed_os "windows")
+        set(OS_UP WINDOWS)
+    else()
+        message(FATAL_ERROR "Error: unknown OS for target ${os}")
+    endif()
+
+    file(APPEND ${CMAKE_BINARY_DIR}/bitcode_libs_generated.cpp
+      "static BitcodeLib ${name}(\"${name}.bc\", ISPCTarget::generic_${target}, TargetOS::${fixed_os}, Arch::${arch});\n")
+    set(target "generic-${target}")
+
+    set(cpp ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/${name}.cpp)
+    set(bc ${BITCODE_FOLDER}/${name}.bc)
+
+    if ("${fixed_os}" STREQUAL "linux" AND NOT ISPC_LINUX_TARGET)
+        # If ISPC_LINUX_TARGET is disabled then we can't run ispc-slim with
+        # --target-os=linux to generate generic target bitcode. So we need pass
+        # the target-os that is supported by ispc-slim.
+        if (APPLE)
+            set(fixed_os "macos")
+        elseif (CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+            set(fixed_os "freebsd")
+        else ()
+            message(FATAL_ERROR "Error: unknown OS for target ${fixed_os}")
+        endif()
+    endif()
+
+    add_custom_command(
+        OUTPUT ${bc}
+        COMMAND ${ispc_name} -I ${include} --enable-llvm-intrinsics --nostdlib --gen-stdlib --target=${target} --arch=${arch} --target-os=${fixed_os}  --emit-llvm -o ${bc} ${input_ispc} -DBUILD_OS=${OS_UP} -DRUNTIME=${bit}
+        DEPENDS ${ispc_name} ${input_ispc}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    )
+
+    add_custom_command(
+        OUTPUT ${cpp}
+        COMMAND ${Python3_EXECUTABLE} ${BITCODE2CPP} ${bc} --type=ispc-target --runtime=${bit} --os=${OS_UP} --arch=${arch} ${cpp}
+        DEPENDS ${bc} ${BITCODE2CPP}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    )
+
+    list(APPEND GENERIC_TARGET_BC_FILE ${bc})
+    set(GENERIC_TARGET_BC_FILE ${GENERIC_TARGET_BC_FILE} PARENT_SCOPE)
+
+    list(APPEND GENERIC_TARGET_CPP_FILE ${cpp})
+    set(GENERIC_TARGET_CPP_FILE ${GENERIC_TARGET_CPP_FILE} PARENT_SCOPE)
+endfunction()
+
+function (generate_generic_builtins ispc_name)
```

Comment:
Same - I would really appreciate the comments for the function as a whole.

Reply:
I have added some comments.